source/application/AMPIC887DriverServer/AMPIC887Controller.cpp

```diff
@@ -61,10 +114,954 @@ QString AMPIC887Controller::status() const
+{
+	AMGCS2AsyncMoveRelativeCommand* moveRelativeCommand = qobject_cast<AMGCS2AsyncMoveRelativeCommand*>(command);
+
+	if(moveRelativeCommand) {
+
+		AMPIC887AxisCollection axesMoving = moveRelativeCommand->relativeTargetPositions().axes();
+		foreach(AMGCS2::Axis currentAxis, axesMoving) {
+			switch (currentAxis) {
+			case AMGCS2::XAxis:
+				--xMotions_;
+				break;
+			case AMGCS2::YAxis:
+				--yMotions_;
+				break;
+			case AMGCS2::ZAxis:
+				--zMotions_;
+				break;
+			case AMGCS2::UAxis:
+				--uMotions_;
+				break;
+			case AMGCS2::VAxis:
+				--vMotions_;
+				break;
+			case AMGCS2::WAxis:
+				--wMotions_;
+				break;
+			default:
+				break;
+			}
+		}
+	}
+
+	command->deleteLater();
+}
+
+// observing the bahviour of the controller shows that the statuses of the
+// axes when a reference move is under way are all or nothing (i.e. if
+// a reference move is started on X, the referenced status of all axes
+// is set to unreferenced, and the movement status of all axes is set to
+// moving).
+void AMPIC887Controller::onAsyncReferenceMoveStarted(AMGCS2AsyncCommand *)
+{
+	currentPositionRefreshRequired_ = true;
+	++xMotions_;
+	++yMotions_;
+	++zMotions_;
+	++uMotions_;
+	++vMotions_;
+	++wMotions_;
+
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::XAxis, false);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::YAxis, false);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::ZAxis, false);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::UAxis, false);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::VAxis, false);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::WAxis, false);
+
+	controllerState_->hexapodState()->setTargetPosition(AMGCS2::XAxis, 0);
+	controllerState_->hexapodState()->setTargetPosition(AMGCS2::YAxis, 0);
+	controllerState_->hexapodState()->setTargetPosition(AMGCS2::ZAxis, 0);
+	controllerState_->hexapodState()->setTargetPosition(AMGCS2::UAxis, 0);
+	controllerState_->hexapodState()->setTargetPosition(AMGCS2::VAxis, 0);
+	controllerState_->hexapodState()->setTargetPosition(AMGCS2::WAxis, 0);
+}
+
+void AMPIC887Controller::onAsyncReferenceMoveSucceeded(AMGCS2AsyncCommand *command)
+{
+	--xMotions_;
+	--yMotions_;
+	--zMotions_;
+	--uMotions_;
+	--vMotions_;
+	--wMotions_;
+
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::XAxis, true);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::YAxis, true);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::ZAxis, true);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::UAxis, true);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::VAxis, true);
+	controllerState_->hexapodState()->setReferencedState(AMGCS2::WAxis, true);
+
+	command->deleteLater();
+}
+
+void AMPIC887Controller::onAsyncReferenceMoveFailed(AMGCS2AsyncCommand *command)
+{
+	--xMotions_;
+	--yMotions_;
+	--zMotions_;
+	--uMotions_;
+	--vMotions_;
+	--wMotions_;
+
```

Comment:
All the referenced states of the axes are taken care of by setting their referenced states to false when the reference move begins. 

I've added a call to updateStateOnStop() which refreshes the current position values, and sets the target position to the current position.